### ISSUE TITLE
Follow-up to updating the Krumo library

### ISSUE THREAD

**Issue by jenlampton** [backdrop-contrib/devel#135]
This is a follow-up to https://github.com/backdrop-contrib/devel/issues/102

> There's obviously more work to do here, like deciding what removed customisations are still needed (and working out how to add them back in, preferably without hacking Krumo again).

From that issue (thank you @BWPanda)

> It seems Devel has hacked Krumo and changed a bunch of stuff in Krumo's files - including adding the addCssJs() function.

> Here's a diff I made of Krumo v0.2.1a downloaded from their GitHub repo, and Devel's copy (diff -uprwiNEB --ignore-file-name-case krumo-0.2.1a/ krumo/): [krumo.diff.txt](https://github.com/backdrop-contrib/devel/files/5473715/krumo.diff.txt)

**Comment by ghost** [comment]
One of the main differences I've noticed between this new, clean version and the old, hacked version is the calling function listed at the bottom of the Krumo output.

My limited testing shows it always seems to say Devel is the caller. One of the hacks we had before was to filter the backtrace by skipping functions with 'devel' in the name, obviously for this reason.

Not sure how to address this now. Maybe a feature request to Krumo that adds the ability to specify additional backtrace skips...? (it already skips 'krumo' functions).

**Comment by jenlampton** [comment]
> Not sure how to address this now. Maybe a feature request to Krumo that adds the ability to specify additional backtrace skips...? (it already skips 'krumo' functions).

That sounds like a great idea. Maybe someone else has opened a similar issue in their queue before?

**Comment by docwilmot** [comment]
I tried the latest dev and did a `dpm()` on a views row `$fields` variable and keep getting `Error: Maximum function nesting level of '256' reached, aborting! in is_callable() (line 1109 of modules\devel\lib\krumo\class.krumo.php)`, which doesnt happen if I revert to previous version

**Comment by argiepiano** [comment]
I can confirm this problem dumping a view structure. To reproduce:

```php
$v = views_get_view('node_admin_content');
$v->set_display('page');
$v->execute();
dpm($v);
```
You get a 500 error in the browser's console, and the nesting error posted above  by @docwilmot.

**Comment by docwilmot** [comment]
Got the same nesting error yesterday on a FormAPI $variables array. I wonder if this is a more general issue than just Views. 

It's much much faster though.

**Comment by argiepiano** [comment]
Agreed. Perhaps some of the krumo customizations that the upgrade removed were necessary. I'm wondering if it may be worth looking at the current D7 version, which is fast and works well with views and other complex structures.

**Comment by jenlampton** [comment]
I would love some help with this. Last I looked at the Drupal 7 version it was still using the older customized version of the Krumo library. I'm going to roll back this change until we can resolve some of these issues, and make a new release with the other fixes first.

**Comment by jenlampton** [comment]
Okay, here is a branch with the krumo update and the linux line-endings fix @quicksketch added: https://github.com/backdrop-contrib/devel/pull/153

We should keep working on this here and can merge once enough people feel it's ready.

**Comment by yorkshire-pudding** [comment]
@jenlampton @docwilmot @argiepiano - has any consideration been given to using [Kint](https://kint-php.github.io/kint/) instead of Krumo?
Specifically, it has this listed in the Advanced page which might resolve some of the issues listed above:

> `BlacklistPlugin`
> Lets you hide information you don’t want dumped, and improve performance by skipping heavy values.

While it has composer installation, it also seems to be very easy to include without composer and the instructions always give both ways.
